File: androidApp/src/main/kotlin/it/mensa/app/features/deals/DealListViewModel.kt
package it.mensa.app.features.deals

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import it.mensa.app.support.koinAccess
import it.mensa.shared.model.DealModel
import it.mensa.shared.model.UserModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

// ─── UI state ────────────────────────────────────────────────────────────────

data class DealListUiState(
    val allDeals: List<DealModel> = emptyList(),
    val refreshing: Boolean = false,
    val error: String? = null,
    val currentUser: UserModel? = null,
    val selectedSector: String? = null,
    val searchText: String = "",
) {
    val sectors: List<String>
        get() = allDeals
            .map { it.commercialSector }
            .filter { it.isNotBlank() }
            .toSet()
            .sorted()

    val filteredDeals: List<DealModel>
        get() {
            val query = searchText.trim().lowercase()
            return allDeals.filter { d ->
                if (selectedSector != null && d.commercialSector != selectedSector) return@filter false
                if (query.isEmpty()) return@filter true
                if (d.name.lowercase().contains(query)) return@filter true
                if (d.details?.lowercase()?.contains(query) == true) return@filter true
                if (d.commercialSector.lowercase().contains(query)) return@filter true
                false
            }
        }

    val isFilterActive: Boolean
        get() = selectedSector != null || searchText.isNotBlank()

    val canAddDeal: Boolean
        get() {
            val user = currentUser ?: return false
            val allowed = setOf("super", "admin", "deals", "deals_admin")
            return user.powers.any { it in allowed }
        }
}

// ─── ViewModel ───────────────────────────────────────────────────────────────

class DealListViewModel : ViewModel() {

    private val koin = koinAccess()

    private val _state = MutableStateFlow(DealListUiState())
    val state: StateFlow<DealListUiState> = _state.asStateFlow()

    init {
        // Observe deals cache flow
        viewModelScope.launch {
            koin.deals.observeAll().collect { deals ->
                _state.update { it.copy(allDeals = deals) }
            }
        }
        // Observe current user flow
        viewModelScope.launch {
            koin.auth.currentUser.collect { user ->
                _state.update { it.copy(currentUser = user) }
            }
        }
        // Initial background refresh
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            _state.update { it.copy(refreshing = true, error = null) }
            try {
                koin.deals.refresh()
            } catch (e: Exception) {
                _state.update { it.copy(error = e.message ?: "Errore sconosciuto") }
            } finally {
                _state.update { it.copy(refreshing = false) }
            }
        }
    }

    fun setSearchText(text: String) {
        _state.update { it.copy(searchText = text) }
    }

    fun setSelectedSector(sector: String?) {
        _state.update { it.copy(selectedSector = sector) }
    }

    fun clearError() {
        _state.update { it.copy(error = null) }
    }
}

File: iosApp/iosApp/Features/Deals/DealListViewModel.swift
import SwiftUI
import Shared

@MainActor
@Observable
final class DealListViewModel {
    /// Raw, unfiltered list straight from the cache flow.
    var allDeals: [DealModel] = []

    /// True only when we have nothing on screen yet AND a refresh is in flight.
    /// Cache-first: if `allDeals` is non-empty we never show a spinner.
    var refreshing = false
    var error: String?

    /// Active sector chip (`nil` = "all").
    var selectedSector: String?
    var searchText: String = ""

    /// Current authenticated user — letto sincrono dall'auth (session-stable,
    /// cambia solo a login/logout e in quel caso la view viene smontata).
    var currentUser: UserModel? {
        koin.auth.currentUser.value as? UserModel
    }

    private var dealsSub: Closeable?

    // MARK: - Lifecycle

    func start() {
        dealsSub?.close()
        dealsSub = FlowBridgeKt.subscribe(
            flow: koin.deals.observeAll(),
            onEach: { [weak self] value in
                Task { @MainActor in
                    self?.allDeals = (value as? [DealModel]) ?? []
                }
            },
            onError: { _ in }
        )

        Task { await refresh() }
    }

    func stop() {
        dealsSub?.close()
        dealsSub = nil
    }

    func refresh() async {
        refreshing = true
        defer { refreshing = false }
        do {
            try await koin.deals.refresh(filter: nil, sort: "created")
        } catch {
            self.error = (error as NSError).localizedDescription
        }
    }

    // MARK: - Derived

    /// All distinct, non-empty sectors present in the dataset, alphabetically.
    var sectors: [String] {
        let set = Set(allDeals.map { $0.commercialSector }.filter { !$0.isEmpty })
        return set.sorted()
    }

    /// Filtered by sector + search.
    var filteredDeals: [DealModel] {
        let trimmedQuery = searchText
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
        return allDeals.filter { d in
            if let sector = selectedSector, d.commercialSector != sector { return false }
            guard !trimmedQuery.isEmpty else { return true }
            if d.name.lowercased().contains(trimmedQuery) { return true }
            if let det = d.details?.lowercased(), det.contains(trimmedQuery) { return true }
            if d.commercialSector.lowercased().contains(trimmedQuery) { return true }
            return false
        }
    }

    /// Empty-state copy depends on whether we filtered ourselves into a corner
    /// or really have no data.
    var isFilterActive: Bool {
        selectedSector != nil ||
            !searchText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    var canAddDeal: Bool {
        guard let user = currentUser else { return false }
        // Flutter app uses 'super' / 'deals_admin' / 'admin' interchangeably.
        let allowed: Set<String> = ["super", "admin", "deals_admin"]
        return !user.powers.filter { allowed.contains($0) }.isEmpty
    }
}
